src/shell/input.c

```diff
@@ -0,0 +1,710 @@
+    if (!(txn = flux_kvs_txn_create ()))
+        goto error;
+    if (flux_kvs_txn_put (txn, FLUX_KVS_APPEND, "input", entrystr) < 0)
+        goto error;
+    if (!(f = flux_kvs_commit (in->shell->h, NULL, 0, txn)))
+        goto error;
+    if (flux_future_then (f, -1, shell_input_put_kvs_completion, in) < 0)
+        goto error;
+    if (flux_shell_add_completion_ref (in->shell, "input.kvs") < 0) {
+        log_err ("flux_shell_remove_completion_ref");
+        goto error;
+    }
+    /* f memory responsibility of shell_input_put_kvs_completion()
+     * callback */
+    f = NULL;
+    rc = 0;
+ error:
+    saved_errno = errno;
+    flux_kvs_txn_destroy (txn);
+    free (entrystr);
+    json_decref (entry);
+    flux_future_destroy (f);
+    errno = saved_errno;
+    return rc;
+}
+
+/* Convert 'iodecode' object to an valid RFC 24 data event.
+ * N.B. the iodecode object is a valid "context" for the event.
+ */
+static void shell_input_stdin_cb (flux_t *h,
+                                  flux_msg_handler_t *mh,
+                                  const flux_msg_t *msg,
+                                  void *arg)
+{
+    struct shell_input *in = arg;
+    bool eof = false;
+    json_t *o;
+
+    if (flux_request_unpack (msg, NULL, "o", &o) < 0)
+        goto error;
+    if (iodecode (o, NULL, NULL, NULL, NULL, &eof) < 0)
+        goto error;
+    if (shell_svc_allowed (in->shell->svc, msg) < 0)
+        goto error;
+    if (shell_input_put_kvs (in, o) < 0)
+        goto error;
+    if (eof)
+        flux_msg_handler_stop (mh);
+    if (flux_respond (in->shell->h, msg, NULL) < 0)
+        log_err ("flux_respond");
+    return;
+error:
+    if (flux_respond_error (in->shell->h, msg, errno, NULL) < 0)
+        log_err ("flux_respond");
+}
+
+static void shell_input_type_file_init (struct shell_input *in)
+{
+    struct shell_input_type_file *fp = &(in->stdin_file);
+    fp->fd = -1;
+}
+
+static int shell_input_parse_type (struct shell_input *in)
+{
+    const char *typestr = NULL;
+    int ret;
+
+    if ((ret = flux_shell_getopt_unpack (in->shell, "input",
+                                         "{s?:{s?:s}}",
+                                         "stdin", "type", &typestr)) < 0)
+        return -1;
+
+    if (!ret || !typestr)
+        return 0;
+
+    if (!strcmp (typestr, "service"))
+        in->stdin_type = FLUX_INPUT_TYPE_SERVICE;
+    if (!strcmp (typestr, "file")) {
+        struct shell_input_type_file *fp = &(in->stdin_file);
+
+        in->stdin_type = FLUX_INPUT_TYPE_FILE;
+
+        if (flux_shell_getopt_unpack (in->shell, "input",
+                                      "{s:{s?:s}}",
+                                      "stdin", "path", &(fp->path)) < 0)
+            return -1;
+
+        if (fp->path == NULL) {
+            log_msg ("path for stdin file input not specified");
+            return -1;
+        }
+    }
+    else {
+        log_msg ("invalid input type specified '%s'", typestr);
+        return -1;
```

Comment:
Maybe throw in a test to cover invalid type (at least codecov showed this case wasn't covered)